fe/fe-core/src/main/java/com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java

```diff
@@ -315,6 +300,59 @@ private RefreshJobStatus doMvRefresh(TaskRunContext context, MaterializedViewMet
         return RefreshJobStatus.SUCCESS;
     }
 
+    /**
+     * Prepare the statement and plan for mv refreshing, considering the partitions of ref table
+     */
+    private InsertStmt prepareRefreshPlan(Map<Table, Set<String>> refTableRefreshPartitions,
+                                          Set<String> mvToRefreshedPartitions,
+                                          Map<String, Set<String>> refTablePartitionNames) throws AnalysisException {
+        // 1. Prepare context
+        ConnectContext ctx = mvContext.getCtx();
+        ctx.getAuditEventBuilder().reset();
+        ctx.getAuditEventBuilder()
+                .setTimestamp(System.currentTimeMillis())
+                .setClientIp(mvContext.getRemoteIp())
+                .setUser(ctx.getQualifiedUser())
+                .setDb(ctx.getDatabase());
+        ctx.setThreadLocalInfo();
+
+        // 2. Prepare variables
+        changeDefaultConnectContextIfNeeded(ctx);
+
+        // 3. AST
+        InsertStmt insertStmt = generateInsertAst(mvToRefreshedPartitions, materializedView, ctx);
+
+        // 4. Analyze and prepare partition
+        Map<String, Database> dbs = AnalyzerUtils.collectAllDatabase(ctx, insertStmt);
+        Locker locker = new Locker();
+        ExecPlan execPlan = null;
+        try {
+            StatementPlanner.lock(locker, dbs);
+
+            insertStmt =
+                    analyzeInsertStmt(insertStmt, mvToRefreshedPartitions, refTablePartitionNames, materializedView);
+            execPlan = StatementPlanner.plan(insertStmt, ctx);
```

Comment:
`StatementPlanner.plan` will acquire locks itself. no need use lock to protect?

Reply:
yes, but we cannot release the lock and acquire again, which cannot rule out concurrent modification before planning. 

Meanwhile, the ReentrantLock is reentrant, the Planner could acquire the lock again.